Add `SVGImageElement.prototype.decode()` support

Ahmad-S792 opened the pull request:
#### f2adb8467c81d2df25e8fa651a54d6d9a00b9338
<pre>
Add `SVGImageElement.prototype.decode()` support

<a href="https://bugs.webkit.org/show_bug.cgi?id=279254">https://bugs.webkit.org/show_bug.cgi?id=279254</a>
<a href="https://rdar.apple.com/135404215">rdar://135404215</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

It extends `decode()` support from HTML Image Element as defined in web
specification [1], to SVG Image Element.

[1] <a href="https://html.spec.whatwg.org/multipage/embedded-content.html#dom-img-decoding">https://html.spec.whatwg.org/multipage/embedded-content.html#dom-img-decoding</a>

Credits to Said Abou-Hallawa for test case guidance.

* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::decode):
* Source/WebCore/svg/SVGImageElement.h:
* Source/WebCore/svg/SVGImageElement.idl:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes-svg.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-svg.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach-svg.tentative-expected.txt:
* LayoutTests/fast/images/decode-static-svg-image-resolve-expected.html:
* LayoutTests/fast/images/decode-static-svg-image-resolve.html:

Canonical link: <a href="https://commits.webkit.org/283643@main">https://commits.webkit.org/283643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26e5b87e882e3b6a6b6235c6e9a10765370b6a2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19532 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/18043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69028 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17824 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/70945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/18043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69977 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/57894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/39254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/15286 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/16397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/61172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15627 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10867 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/72646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57951 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14776 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/8927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/43169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44352 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42912 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes-svg.tentative-expected.txt:
```diff
@@ -1,8 +1,8 @@
-FAIL SVGImageElement.prototype.decode(), href mutation tests. href changes fail decode; following good decode succeeds. img.decode is not a function. (In 'img.decode()', 'img.decode' is undefined)
-FAIL SVGImageElement.prototype.decode(), href mutation tests. xlink:href changes fail decode; following bad decode fails. img.decode is not a function. (In 'img.decode()', 'img.decode' is undefined)
-FAIL SVGImageElement.prototype.decode(), href mutation tests. href changes fail decode; following bad decode fails. img.decode is not a function. (In 'img.decode()', 'img.decode' is undefined)
+FAIL SVGImageElement.prototype.decode(), href mutation tests. xlink:href changes fail decode. assert_unreached: Should have rejected: undefined Reached unreachable code
+FAIL SVGImageElement.prototype.decode(), href mutation tests. href changes fail decode. assert_unreached: Should have rejected: undefined Reached unreachable code
+FAIL SVGImageElement.prototype.decode(), href mutation tests. xlink:href changes fail decode; following good decode succeeds. assert_unreached: Should have rejected: undefined Reached unreachable code
+FAIL SVGImageElement.prototype.decode(), href mutation tests. href changes fail decode; following good decode succeeds. assert_unreached: Should have rejected: undefined Reached unreachable code
```
What is the problem here? Are these tests wrong?

Ahmad-S792 replied:
Honestly, I haven't investigated much because I was thinking of it as stub implementation.

annevk replied:
If it's a stub implementation it needs to be behind a preference. Definitely if you're not sure about it.

Ahmad-S792 replied:
I just closed PR for future for someone to take care of it. :-)

shallawa replied:
These tests are failing because changing the image source should cause the decoding promise to be rejected. But currently it is resolved.

I think this change is fine. With it, `SVGImageElement.decode()` will be working as the `HTMLImageElement.decode()`. I am almost sure that I can reproduce the same errors with `HTMLImageElement`.

So my suggestions are:

1. Reopen this PR and please add at least one layout test for SVGImageElement. You can add a test similar to fast/images/decode-static-image-resolve.html.
2. File a new bug for the remaining issues, like: "Changing the image source does not reject the decode() promise". This bug should cover the `HTMLImageElement` and `SVGImageElement`.

Ahmad-S792 replied:
@shallawa - Thanks for your input. I have added test as suggested and opened new bug here - https://bugs.webkit.org/show_bug.cgi?id=279623

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve.html:
```diff
@@ -0,0 +1,24 @@
+        if (window.testRunner)
+            testRunner.waitUntilDone();
+        var image = new Image;
+        image.src = "resources/green-400x400.svg";
```
This test does not test the `SVGImageElement`. `new Image` creates an `HTMLImageElement`. Decoding the `green-400x400.svg` will in fact do nothing. It will just resolve the promise.

I attached a test case to https://bugs.webkit.org/show_bug.cgi?id=279254.

Ahmad-S792 replied:
@shallawa - thanks! I only changed two bits, removed `paragraph` and made it as inline comment in CSS and change image reference in your test case. Plus accordingly updated test case expectation file to have two green rects.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve.html:
```diff
@@ -0,0 +1,37 @@
+<style>
+/* This tests passes if two green rectangles are drawn. */
```
This comment looks like a debugging comment. Usually we put this message in `<p>...</p>` or `<h3>...</h3>` to make it easy to know what is expected from this file once it is displayed.  Some tests put this message in `<title>...</title>`.

Ahmad-S792 replied:
Added `<p>...</p>`.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve-expected.html:
```diff
@@ -0,0 +1,22 @@
+    }
+</style>
+<body>
+    <div class="container">
```
What is `container`? Do we really need this `<div>`?

Ahmad-S792 replied:
Removed it.

shallawa commented on Source/WebCore/svg/SVGImageElement.cpp:
```diff
@@ -219,4 +219,14 @@ void SVGImageElement::didMoveToNewDocument(Document& oldDocument, Document& newD
+size_t SVGImageElement::pendingDecodePromisesCountForTesting() const
+{
+    return m_imageLoader.pendingDecodePromisesCountForTesting();
+}
```
This function is not used in the code right now. Please remove it.

shallawa commented on Source/WebCore/svg/SVGImageElement.h:
```diff
@@ -55,6 +55,9 @@ class SVGImageElement final : public SVGGraphicsElement, public SVGURIReference
 
     using PropertyRegistry = SVGPropertyOwnerRegistry<SVGImageElement, SVGGraphicsElement, SVGURIReference>;
 
+    WEBCORE_EXPORT void decode(Ref<DeferredPromise>&&);
```
This function is not used outside WebCore. Please remove `WEBCORE_EXPORT`.

Ahmad-S792 replied:
Removed `WEBCORE_EXPORT`.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve-expected.html:
```diff
@@ -0,0 +1,22 @@
+    .rectangles {
+        display: inline-block;
+    }
+    .rectangle {
```
This can be named `square` or `box`.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve-expected.html:
```diff
@@ -0,0 +1,20 @@
+    }
+</style>
+<body>
+<p>This tests passes if two green rectangles are drawn.</p>
```
Please indent.

Ahmad-S792 replied:
Done.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve.html:
```diff
@@ -0,0 +1,38 @@
+<style>
+/* This tests passes if two green rectangles are drawn. */
```
Please remove this comment. It is already in the body as a `<p>` element.

Ahmad-S792 replied:
Done.

shallawa commented on LayoutTests/fast/images/decode-static-svg-image-resolve.html:
```diff
@@ -0,0 +1,38 @@
+    }
+</style>
+<body>
+<p>This tests passes if two green rectangles are drawn.</p>
```
Please indent.

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
Add `SVGImageElement.prototype.decode()` support

https://bugs.webkit.org/show_bug.cgi?id=279254
rdar://135404215

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

It extends `decode()` support from HTML Image Element as defined in web
specification [1], to SVG Image Element.

[1] https://html.spec.whatwg.org/multipage/embedded-content.html#dom-img-decoding

Credits to Said Abou-Hallawa for test case guidance.

* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::decode):
* Source/WebCore/svg/SVGImageElement.h:
* Source/WebCore/svg/SVGImageElement.idl:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes-svg.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-svg.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach-svg.tentative-expected.txt:
* LayoutTests/fast/images/decode-static-svg-image-resolve-expected.html:
* LayoutTests/fast/images/decode-static-svg-image-resolve.html:

Canonical link: https://commits.webkit.org/283643@main
